File: 043_Group Anagrams.py
"""
Group Anagrams

Group an array of strings into lists of anagrams.
"""

from typing import List
from collections import defaultdict

def group_anagrams(strs: List[str]) -> List[List[str]]:
    mp = defaultdict(list)
    for s in strs:
        key = ''.join(sorted(s))
        mp[key].append(s)
    return list(mp.values())


if __name__ == '__main__':
    print('Groups:', group_anagrams(["eat","tea","tan","ate","nat","bat"]))

# Time Complexity: O(n * k log k)
# Space Complexity: O(n*k)

File: 043_Group Anagrams.cpp
/*
Group Anagrams

Group an array of strings into lists of anagrams.
*/

#include <bits/stdc++.h>
using namespace std;

vector<vector<string>> groupAnagrams(vector<string>& strs) {
    unordered_map<string, vector<string>> mp;
    for (auto &s: strs) {
        string key = s;
        sort(key.begin(), key.end());
        mp[key].push_back(s);
    }
    vector<vector<string>> res;
    for (auto &p: mp) res.push_back(p.second);
    return res;
}

int main() {
    vector<string> s1 = {"eat","tea","tan","ate","nat","bat"};
    auto r1 = groupAnagrams(s1);
    cout << "Groups:" << endl;
    for (auto &g: r1) {
        for (auto &w: g) cout << w << ' ';
        cout << endl;
    }
    return 0;
}

// Time Complexity: O(n * k log k)
// Space Complexity: O(n*k)
